orca: implement intra-segment parallel table scan support

yjhjstz opened the pull request:
Add comprehensive parallel table scan capability to GPORCA optimizer, enabling worker-level parallelism within segments for improved query performance on large table scans.

Key components:
- New CPhysicalParallelTableScan operator and CDistributionSpecWorkerRandom distribution specification for worker-level data distribution
- CXformGet2ParallelTableScan transformation with parallel safety checks (excludes CTEs, dynamic scans, foreign tables, replicated tables, etc.)
- Cost model integration with parallel_setup_cost and efficiency degradation scaling (logarithmic based on worker count)
- DXL serialization/deserialization for CDXLPhysicalParallelTableScan
- Plan translation to PostgreSQL SeqScan nodes with parallel_aware=true
- Rewindability constraints (parallel scans are non-rewindable)
- GUC integration: max_parallel_workers_per_gather controls worker count

Impl https://github.com/apache/cloudberry/discussions/1316

TPCH improved 15%.


### What does this PR do?
<!-- Brief overview of the changes, including any major features or fixes -->

### Type of Change
- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature with breaking changes)
- [ ] Documentation update

### Breaking Changes
<!-- Remove if not applicable. If yes, explain impact and migration path -->

### Test Plan
<!-- How did you test these changes? -->
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] Passed `make installcheck`
- [ ] Passed `make -C src/test installcheck-cbdb-parallel`

### Impact
<!-- Remove sections that don't apply -->
**Performance:**
<!-- Any performance implications? -->

**User-facing changes:**
<!-- Any changes visible to users? -->

**Dependencies:**
<!-- New dependencies or version changes? -->

### Checklist
- [ ] Followed [contribution guide](https://cloudberry.apache.org/contribute/code)
- [ ] Added/updated documentation
- [ ] Reviewed code for security implications
- [ ] Requested review from [cloudberry committers](https://github.com/orgs/apache/teams/cloudberry-committers)

### Additional Context
<!-- Any other information that would help reviewers? Remove if none -->

### CI Skip Instructions
<!--
To skip CI builds, add the appropriate CI skip identifier to your PR title.
The identifier must:
- Be in square brackets []
- Include the word "ci" and either "skip" or "no"
- Only use for documentation-only changes or when absolutely necessary
-->

---
<!-- Join our community:
- Mailing list: [dev@cloudberry.apache.org](https://lists.apache.org/list.html?dev@cloudberry.apache.org) (subscribe: dev-subscribe@cloudberry.apache.org)
- Discussions: https://github.com/apache/cloudberry/discussions -->

yjhjstz committed:
orca: implement intra-segment parallel table scan support

Add comprehensive parallel table scan capability to GPORCA optimizer,
enabling worker-level parallelism within segments for improved query
performance on large table scans.

Key components:
- New CPhysicalParallelTableScan operator and CDistributionSpecWorkerRandom
distribution specification for worker-level data distribution
- CXformGet2ParallelTableScan transformation with parallel safety checks
(excludes CTEs, dynamic scans, foreign tables, replicated tables, etc.)
- Cost model integration with parallel_setup_cost and efficiency degradation
scaling (logarithmic based on worker count)
- DXL serialization/deserialization for CDXLPhysicalParallelTableScan
- Plan translation to PostgreSQL SeqScan nodes with parallel_aware=true
- Rewindability constraints (parallel scans are non-rewindable)
- GUC integration: max_parallel_workers_per_gather controls worker count

my-ship-it commented:
Please add more test cases

my-ship-it commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -6156,9 +6297,11 @@ CTranslatorDXLToPlStmt::TranslatePlanCosts(const CDXLNode *dxlnode, Plan *plan)
+	double total_rows = CostFromStr(costs->GetRowsOutStr());
+	double rows_per_segment = total_rows / current_slice->numsegments;
+	
+	plan->plan_rows = ceil(rows_per_segment);
```
Why change codes here?

my-ship-it commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -2410,6 +2416,120 @@ CCostModelGPDB::CostScan(CMemoryPool *,	 // mp
+
+	// Efficiency decreases logarithmically with more workers
+	// Formula: efficiency = 1 / (1 + 0.1 * log2(workers))
+	// This gives: 2 workers = 0.91, 4 workers = 0.83, 8 workers = 0.77
```
Where is the formula from?

my-ship-it commented on src/backend/gporca/libgpopt/include/gpopt/base/CRewindabilitySpec.h:
```diff
@@ -108,6 +108,13 @@ class CRewindabilitySpec : public CPropSpec
 		EmhtSentinel
 	};
 
+	enum EBarrierHazardType
```
Why we declare enum in RewindabilitySpec?

my-ship-it commented on src/backend/gporca/libgpopt/src/operators/CPhysicalParallelTableScan.cpp:
```diff
@@ -0,0 +1,284 @@
+	{
+		// Create worker-level random distribution using the table's distribution as base
+		// The base CPhysicalScan already sets up m_pds from the table descriptor
+		m_pdsWorkerDistribution = CDistributionSpecWorkerRandom::PdsCreateWorkerRandom(mp, ulParallelWorkers, m_pds);
```
It looks like a composite spec, e.g, original spec + parallel workers. Is WorkerRandom accurate here?

yjhjstz replied:
anyway,  it is random.

yjhjstz commented on src/backend/gpopt/gpdbwrappers.cpp:
```diff
@@ -2706,4 +2726,36 @@ gpdb::TestexprIsHashable(Node *testexpr, List *param_ids)
+		if (!enable_parallel)
+			return false;
+
+		if (IS_SINGLENODE())
```
orca not support single node mode yet.

yjhjstz commented on src/backend/gporca/libgpopt/src/operators/CPhysicalParallelTableScan.cpp:
```diff
@@ -0,0 +1,284 @@
+
+	// Otherwise, derive from the base physical scan
+	// This uses the m_pds member from CPhysicalScan
+	return CPhysicalScan::PdsDerive(mp, exprhdl);
```
will not happen

my-ship-it commented on src/test/regress/GNUmakefile:
```diff
@@ -223,6 +223,12 @@ installcheck-cbdb-parallel: all tablespace-setup twophase_pqexecparams
+installcheck-orca-parallel: all tablespace-setup twophase_pqexecparams
+	( \
+	export PGOPTIONS='-c optimizer=on -c enable_parallel=true -c min_parallel_table_scan_size=0 -c min_parallel_index_scan_size=0 -c parallel_setup_cost=0'; \
+	$(pg_regress_installcheck) $(REGRESS_OPTS) --exclude-file=$(srcdir)/excluded_tests.conf --schedule=$(srcdir)/parallel_schedule --schedule=$(srcdir)/greenplum_schedule --max-connections=4 --ignore-plans $(EXTRA_TESTS) \
```
why ignore-plans here?

my-ship-it reviewed (approved):
LGTM

avamingli commented:
Add some cases to  test the plan?

yjhjstz commented:
> Add some cases to test the plan?

maybe after impl parallel hash join .

yjhjstz committed:
PAX stats test: wait for seq_tup_read before exiting wait_for_stats()

In Cloudberry's MPP architecture, segment stats are delivered
asynchronously to the coordinator. The seq_scan counter can be
registered before seq_tup_read arrives from segments, causing
wait_for_stats() to exit prematurely and the subsequent assertion
to fail intermittently in the pax-ic-good-opt-off CI job.

Add an explicit wait condition (updated6) for seq_tup_read reaching
the expected value, and update the comment to reflect Cloudberry's
segment-level async stats delivery rather than parallel workers.

zhangwenchao-123 commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -2416,6 +2422,120 @@ CCostModelGPDB::CostScan(CMemoryPool *,	 // mp
+	CDouble dWorkerStartupCost = GetWorkerStartupCost(pcmgpdb, ulWorkers);
+
+	// Total cost
+	return CCost(pci->NumRebinds() * (dParallelScanCost + dWorkerStartupCost));
```
Whether this CCost(pci->NumRebinds()) * dParallelScanCost + dWorkerStartupCost;?

yjhjstz replied:
fixed.

zhangwenchao-123 commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -2489,11 +2609,15 @@ CCostModelGPDB::Cost(
 
+		case COperator::EopPhysicalParallelTableScan:
+		{
+			return CostParallelTableScan(m_mp, exprhdl, this, pci);
```
How about we move the CostParallelTableScan logic into CostScan?

yjhjstz replied:
better not.

zhangwenchao-123 commented on src/backend/gporca/libgpopt/src/base/CCostContext.cpp:
```diff
@@ -505,6 +505,7 @@ CCostContext::FBetterThan(const CCostContext *pcc) const
 	if (CDistributionSpec::EdtHashed == Pdpplan()->Pds()->Edt() &&
 		CDistributionSpec::EdtRandom == pcc->Pdpplan()->Pds()->Edt())
 	{
+		// FIXME:
```
Here forget or just mark?

yjhjstz replied:
removed.

zhangwenchao-123 commented on src/backend/gporca/libgpopt/src/xforms/CXformGet2ParallelTableScan.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	// Determine parallel workers degree
+	// Priority: table-level parallel_workers setting > GUC max_parallel_workers_per_gather > default
+	ULONG ulParallelWorkers = 2;  // default
```
Add a macros for parallel workers 2？

yjhjstz replied:
add a macros.

zhangwenchao-123 commented:
LGTM.

yjhjstz committed:
orca: add parallel scan test
